### PR TITLE
Add command to delete profiles from a token

### PR DESCRIPTION
**Pull request by ZoltanFridrich**
Adds the ability to delete profile objects from given PKCS#11 token.

$ p11-kit delete-profile --profile public-certificates-token 'pkcs11:token=MyToken'

### CODE REVIEW

File: p11-kit/delete-profile.c [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+			if (profile == CKP_INVALID_ID)
+				profile = strtol (optarg, NULL, 16);
+			if (profile == CKP_INVALID_ID) {
+				p11_message (_("failed to convert profile argument: %s"), optarg);
+				goto cleanup;
+			}
+
+			if (n_profiles >= alloc_profiles) {
+				alloc_profiles += 4;
+				tmp = realloc (profiles, alloc_profiles * sizeof (CK_PROFILE_ID));
```
Comment (ueno) [inline p11-kit/delete-profile.c:191]: Would it be useful to support multiple profiles specified at once, given the user could always call `p11-kit delete-profile` multiple times from scripts / terminal? I would have it always take a single profile, following the "Do One Thing and Do It Well" [principle](https://en.wikipedia.org/wiki/Unix_philosophy#Do_One_Thing_and_Do_It_Well).

File: p11-kit/delete-profile.c [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+	argc -= optind;
+	argv += optind;
+
+	if (argc != 1) {
+		p11_tool_usage (usages, options);
+		goto cleanup;
+	}
+
+	ret = delete_profile (*argv, profiles, n_profiles);
```
Comment (ueno) [inline p11-kit/delete-profile.c:217]: Should we check at least one profile is specified?

**Review by ueno — COMMENTED** [review]
Can we add tests in a similar manner to list-profiles?

File: p11-kit/delete-profile.c [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+			p11_kit_be_quiet ();
+			break;
+		case opt_help:
+			p11_tool_usage (usages, options);
+			ret = 0;
+			goto cleanup;
+		case opt_profile:
+			profile = p11_constant_resolve (profile_nicks, optarg);
+			if (profile == CKP_INVALID_ID)
+				profile = strtol (optarg, NULL, 16);
```
Comment (ueno) [inline p11-kit/delete-profile.c:183]: I would suggest relying on `strtol(..., 0)`, so the user can specify the profile id in both hexadecimal and decimal format.

Comment (ZoltanFridrich) [reply p11-kit/delete-profile.c:183]: ok

File: p11-kit/delete-profile.c (hunk 2 above)
Comment (ZoltanFridrich) [reply p11-kit/delete-profile.c:217]: probably yes.

File: p11-kit/delete-profile.c (hunk 1 above)
Comment (ZoltanFridrich) [reply p11-kit/delete-profile.c:191]: In such case why do we even have an option --profile when the option is mandatory and you cant specify more then one. The command might as well just take two arguments <profile> and <token>

**Comment by ZoltanFridrich** [comment]
> Can we add tests in a similar manner to list-profiles?

I don't think so. The mock modules are loaded when the command is executed and the artificial profile is created. The problem is that I cant check with list-profiles whether the profile has been deleted because that command loads the mock modules again and therefore creates the artificial profile again.

**Comment by ueno** [comment]
> I don't think so. The mock modules are loaded when the command is executed and the artificial profile is created. The problem is that I cant check with list-profiles whether the profile has been deleted because that command loads the mock modules again and therefore creates the artificial profile again.

Yeah, I also can't think of any easy way to test it so let's not block on this. Maybe we could have a special mock module backed by a file to load/store profile objects, or perhaps it might be sufficient to simply check `C_DestroyObject` is called with `log-calls: yes` in the module configuration.

**Comment by ueno** [comment]
> I don't think so. The mock modules are loaded when the command is executed and the artificial profile is created. The problem is that I cant check with list-profiles whether the profile has been deleted because that command loads the mock modules again and therefore creates the artificial profile again.

Yeah, I also can't think of any easy way to test it so let's not block on this. Maybe we could have a special mock module backed by a file to load/store profile objects, or perhaps it might be sufficient to simply check `C_DestroyObject` is called with `log-calls: yes` in the module configuration.

**Pull request merged**